storage/innobase/srv/srv0start.cc

```diff
@@ -1339,47 +1301,80 @@ dberr_t srv_start(bool create_new_db)
-			if (flushed_lsn < lsn_t(1000)) {
-				ib::error()
-					<< "Cannot create log file because"
-					" data files are corrupt or the"
-					" database was not shut down cleanly"
-					" after creating the data files.";
-				return srv_init_abort(DB_ERROR);
-			}
+		if (!create_new_log) {
+			srv_log_file_found = log_file_found;
 
-			srv_log_file_size = srv_log_file_size_requested;
+			log_sys.log.open_file(get_log_file_path());
 
-			err = create_log_file(false, flushed_lsn, logfile0);
+			log_sys.log.create();
 
-			if (err == DB_SUCCESS) {
-				err = create_log_file_rename(flushed_lsn,
-							     logfile0);
+			if (!log_set_capacity(
+				srv_log_file_size_requested)) {
+				return(srv_init_abort(DB_ERROR));
 			}
 
+			/* Enable checkpoints in the page cleaner. */
+			recv_sys.recovery_on = false;
+
+			err= recv_recovery_read_max_checkpoint();
+
 			if (err != DB_SUCCESS) {
-				return(srv_init_abort(err));
+				return srv_init_abort(err);
 			}
-
-			/* Suppress the message about
-			crash recovery. */
-			flushed_lsn = log_sys.get_lsn();
-			goto file_checked;
 		}
+	}
 
-		srv_log_file_found = log_file_found;
+	/* Open or create the data files. */
+	ulint	sum_of_new_sizes;
 
-		log_sys.log.open_file(get_log_file_path());
+	err = srv_sys_space.open_or_create(
+		false, create_new_db, &sum_of_new_sizes, &flushed_lsn);
 
-		log_sys.log.create();
+	switch (err) {
+	case DB_SUCCESS:
+		break;
+	case DB_CANNOT_OPEN_FILE:
+		ib::error()
+			<< "Could not open or create the system tablespace. If"
+			" you tried to add new data files to the system"
+			" tablespace, and it failed here, you should now"
+			" edit innodb_data_file_path in my.cnf back to what"
+			" it was, and remove the new ibdata files InnoDB"
+			" created in this failed attempt. InnoDB only wrote"
+			" those files full of zeros, but did not yet use"
+			" them in any way. But be careful: do not remove"
+			" old data files which contain your precious data!";
+		/* fall through */
+	default:
+		/* Other errors might come from Datafile::validate_first_page() */
+		return(srv_init_abort(err));
+	}
 
-		if (!log_set_capacity(srv_log_file_size_requested)) {
-			return(srv_init_abort(DB_ERROR));
+	if (!create_new_db && create_new_log) {
+		if (flushed_lsn < lsn_t(1000)) {
+			ib::error()
+				<< "Cannot create log file because"
+				" data files are corrupt or the"
+				" database was not shut down cleanly"
+				" after creating the data files.";
+			return srv_init_abort(DB_ERROR);
+		}
+
+		srv_log_file_size = srv_log_file_size_requested;
+
+		err = create_log_file(false, flushed_lsn, logfile0);
+		if (err == DB_SUCCESS) {
+			err = create_log_file_rename(flushed_lsn, logfile0);
 		}
 
-		/* Enable checkpoints in the page cleaner. */
-		recv_sys.recovery_on = false;
+		if (err != DB_SUCCESS) {
+			return(srv_init_abort(err));
+		}
+
+		/* Suppress the message about
+		crash recovery. */
+		flushed_lsn = log_sys.get_lsn();
+		goto file_checked;
 	}
 
 file_checked:
```

Comment:
The comment after this line needs to be modified. The log file has already been opened.